Pg connect on machines

davissp14 opened the pull request:
Exposes the quick connect command: `DEV=1 fly services postgres connect`. 

<img width="827" alt="Screen Shot 2022-01-19 at 4 56 11 PM" src="https://user-images.githubusercontent.com/423038/150231698-7a0816cd-a3bd-4b5e-b4b6-7c8b4cc87004.png">


cc:// @azazeal  One thing i'm unsure about here is the ssh package.  Would love your thoughts.

azazeal commented on internal/cli/internal/command/services/postgres/connect.go:
```diff
@@ -0,0 +1,93 @@
+func newConnect() (cmd *cobra.Command) {
+	const (
+		long = `
+			Connect to the Postgres console
```
this should't be tabbed; it'll be indented when printed to the output. [Here](https://github.com/superfly/flyctl/blob/79ff1a15d3d3eef7b15b27a084fb84ad6679cbe0/internal/cli/internal/command/apps/apps.go#L13) is a correct example

azazeal commented on internal/cli/internal/command/services/postgres/connect.go:
```diff
@@ -0,0 +1,93 @@
+func runConnect(ctx context.Context) error {
+	client := client.FromContext(ctx).API()
+
+	appName := flag.FirstArg(ctx)
```
If you have a command which expects an app name, we've built helpers for that. You can check how `apps open` behaves, as an example, [here](https://github.com/superfly/flyctl/blob/79ff1a15d3d3eef7b15b27a084fb84ad6679cbe0/internal/cli/internal/command/apps/open.go).

azazeal commented on internal/cli/internal/command/services/postgres/connect.go:
```diff
@@ -0,0 +1,93 @@
+
+	app, err := client.GetApp(ctx, appName)
+	if err != nil {
+		return fmt.Errorf("get app: %w", err)
```
we should probably wrap this common error into its own class. for now please decorate as `apps open` [does](https://github.com/superfly/flyctl/blob/79ff1a15d3d3eef7b15b27a084fb84ad6679cbe0/internal/cli/internal/command/apps/open.go#L52):

```go
if err != nil {
	return fmt.Errorf("failed retrieving app %s: %w", appName, err)
}
```

azazeal commented on internal/cli/internal/command/services/postgres/connect.go:
```diff
@@ -0,0 +1,93 @@
+
+	agentclient, err := agent.Establish(ctx, client)
+	if err != nil {
+		return errors.Wrap(err, "can't establish agent")
```
we don't use `errors.Wrap` but `fmt.Errorf` with `%w`

azazeal commented on internal/cli/internal/command/services/postgres/connect.go:
```diff
@@ -0,0 +1,93 @@
+
+	dialer, err := agentclient.Dialer(ctx, &app.Organization)
+	if err != nil {
+		return fmt.Errorf("failed to build tunnel for %s. %v", app.Organization.Slug, err)
```
I think you added a `.` instead of a `:` before `%w`

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+package ssh
+
```
separate the imports please

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+)
+
+func spin(in, out string) context.CancelFunc {
+	ctx, cancel := context.WithCancel(context.Background())
```
we don't use `context.Background`, in the new CLI we carry over the same context so that we can pickup cancellation.

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+	return cancel
+}
+
+type SSHParams struct {
```
why are you exporting these params and prefixing them with `SSH`?

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+func SSHConnect(p *SSHParams, addr string) error {
+	terminal.Debugf("Fetching certificate for %s\n", addr)
```
we don't use the `terminal` package anymore, `logger.FromContext` is what you're looking for.

davissp14 replied:
This was just a copy and paste from the cmd package to help get me unblocked.

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+
+	cert, err := singleUseSSHCertificate(p.Ctx, p.Org)
+	if err != nil {
+		return fmt.Errorf("create ssh certificate: %w (if you haven't created a key for your org yet, try `flyctl ssh establish`)", err)
```
we also support custom error types which implement `Description() string` for more elaborate error printing. Ping me on slack and I'll guide you through that.

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+
+	pk, err := parsePrivateKey(cert.Key)
+	if err != nil {
+		return errors.Wrap(err, "parse ssh certificate")
```
we don't use `errors.Wrap`, please :)

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+
+	terminal.Debugf("Connection completed.\n", addr)
+
+	if !p.DisableSpinner {
```
I think we need to work on the spinners a bit more

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// // stolen from `mikesmitty`, thanks, you are a mikesmitty and a scholar
+func MarshalED25519PrivateKey(key ed25519.PrivateKey, comment string) []byte {
```
I don't believe you should be exporting this

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+
+// // stolen from `mikesmitty`, thanks, you are a mikesmitty and a scholar
+func MarshalED25519PrivateKey(key ed25519.PrivateKey, comment string) []byte {
+	magic := append([]byte("openssh-key-v1"), 0)
```
this can easily be `magic := []byte("openssh-key-v1\x00")`

davissp14 replied:
Got it, this entire file was copy and pasted over from the `cmd` package.

azazeal commented on internal/cli/internal/command/ssh/ssh_terminal.go:
```diff
@@ -0,0 +1,204 @@
+func parsePrivateKey(key64 string) (ed25519.PrivateKey, error) {
+	pkeys, err := base64.StdEncoding.DecodeString(key64)
+	if err != nil {
+		return nil, fmt.Errorf("API error: can't parse API-provided private key: %w", err)
```
this isn't an API error and you're inside a helper function, you shouldn't be decorating the error here but at the caller.

davissp14 commented:
Appreciate the review!  The connect file should be good to go, i added a TODO for the ssh_terminal file for the time being.  I just moved this over to unblock myself.
